### PR TITLE
Add filter functionality to the select component

### PR DESCRIPTION
**Pull request by efekaptan**
Adds filter functionality to the existing select component. It can be configured with optional `searchable` flag.

### CODE REVIEW

File: src/select/select.tsx [hunk 1]
```diff
@@ -73,23 +86,36 @@ class Select<T> extends React.Component<Props<T>, State> {
...
-                    );
-                })}
-            </ul>
+            <>
+                {this.props.searchable && <div className="select-filter-bar">
+                    <input
+                        type="text"
+                        data-qa={`select--filter-${this.props.id}`}
+                        className="form-control"
+                        placeholder="search"
```
Comment (BenjaminEckardt) [inline src/select/select.tsx]: Please make it configurable via Props. Otherwise there is no way of providing i18n.

File: src/select/select.tsx [hunk 2]
```diff
@@ -34,6 +39,14 @@ class Select<T> extends React.Component<Props<T>, State> {
         this.handleOptionClick = this.handleOptionClick.bind(this);
         this.toggleOptionsList = this.toggleOptionsList.bind(this);
         this.collapseOptionsList = this.collapseOptionsList.bind(this);
+        this.filterOptions = this.filterOptions.bind(this);
+        this.searchInputRef = React.createRef();
+    }
+
+    public componentDidMount() {
+        this.setState({
+            options: this.props.options
```
Comment (BenjaminEckardt) [inline src/select/select.tsx]: Can't you just do that in the constructor instead of assigning empty array first?

File: src/select/select.tsx [hunk 3]
```diff
@@ -34,6 +39,14 @@ class Select<T> extends React.Component<Props<T>, State> {
         this.handleOptionClick = this.handleOptionClick.bind(this);
         this.toggleOptionsList = this.toggleOptionsList.bind(this);
         this.collapseOptionsList = this.collapseOptionsList.bind(this);
+        this.filterOptions = this.filterOptions.bind(this);
```
Comment (BenjaminEckardt) [inline src/select/select.tsx]: I'm picky here but alls the this bindings are currently ordered by length :D

File: src/select/select.tsx [hunk 4]
```diff
@@ -34,6 +39,14 @@ class Select<T> extends React.Component<Props<T>, State> {
...
+        this.searchInputRef = React.createRef();
```
Comment (BenjaminEckardt) [inline src/select/select.tsx]: Please put in before the this bindings

File: src/select/select.tsx (hunk 2 above)
Comment (efekaptan) [reply src/select/select.tsx]: I agree.

File: src/select/select.tsx (hunk 1 above)
Comment (efekaptan) [reply src/select/select.tsx]: Ok, added another optional parameter called `searchPlaceholder`.

File: src/select/select.tsx [hunk 5]
```diff
@@ -129,13 +157,31 @@ class Select<T> extends React.Component<Props<T>, State> {
...
+            this.searchInputRef.current.focus();
+        }
+    }
+
     private isSelected(option: Option<T>) {
         return isEqual(this.props.value, option.value);
     }
+
+    private filterOptions(event: React.ChangeEvent<HTMLInputElement>) {
+        const filteredOptions = this.props.options.filter(option => matches(event.target.value, option.label));
```
Comment (bedakb) [inline src/select/select.tsx]: A bit picky, but usually we'd wrap `option` with parentheses :)

**Pull request merged**